In [1]:
from sklearn.datasets import fetch_california_housing

# Load the housing dataset
housing = fetch_california_housing()

In [3]:
import pandas as pd
X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

In [52]:
#Print the first five rows of the dataset
print("First five rows of the dataset: ")
print(X.head())
#Print the feature names
print("Here are the different features of the dataset: ")
print(X.columns)
#Check for missing values
print("Let's see if there are missing values: ")
print(X.isnull().sum())
print("There are no missing values in our dataset.")



First five rows of the dataset: 
   MedInc  AveBedrms  Longitude
0  8.3252   1.023810    -122.23
1  8.3014   0.971880    -122.22
2  7.2574   1.073446    -122.24
3  5.6431   1.073059    -122.25
4  3.8462   1.081081    -122.25
Here are the different features of the dataset: 
Index(['MedInc', 'AveBedrms', 'Longitude'], dtype='object')
Let's see if there are missing values: 
MedInc       0
AveBedrms    0
Longitude    0
dtype: int64
There are no missing values in our dataset.


In [42]:
#Generate summary statistics
print("Here are our summary statistics: ")
print(X.describe())

print("Here are summary statistics for our target feature:")
print(y.describe())

Here are our summary statistics: 
             MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000  
25% 

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

#Split the dataset into training and test sets (80% training, 20% testing)
X_train_raw, X_test_raw, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Train a linear regression model on the unscaled data
lin_reg_raw = LinearRegression()
lin_reg_raw.fit(X_train_raw, y_train)
#Make predictions on the test set
y_pred_raw = lin_reg_raw.predict(X_test_raw)



In [36]:
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score

#Evaluate model performance using Mean Squared Error (MSE), Root Mean Squared Error (RMSE) and R^2 score
mse_raw = mean_squared_error(y_test, y_pred_raw)
rmse_raw = root_mean_squared_error(y_test, y_pred_raw)
r2_raw = r2_score(y_test, y_pred_raw)

print("Unscaled Data Model Performance Evaluation: ")
print(f"Mean Squared Error: {mse_raw:.2f}")
print(f"Root Mean Squared Error: {rmse_raw:.2f}")
print(f"R² Score: {r2_raw:.2f}")


Unscaled Data Model Performance Evaluation: 
Mean Squared Error: 0.56
Root Mean Squared Error: 0.75
R² Score: 0.58


In [ ]:
#Find the model's coefficients
print("Model Coefficients (Unscaled):")
print(pd.Series(lin_reg_raw.coef_,
                index=X.columns))
print("\nModel Intercept (Unscaled):")
print(pd.Series(lin_reg_raw.intercept_))



Model Coefficients (Unscaled):
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64

Model Intercept (Unscaled):
0   -37.023278
dtype: float64


What does the R^2 score tell us about model performance?
The R^2 score tells us how much the variation between predictions can be accounted for by the linear regression model. The closer a score is to 1, the better the model performs. Therefore, a score of 0.58 that this linear regression model performs moderately well, but not great.
Which features seem to have the strongest impact on predictions based on the model's coefficients?
The features that seem to have the strongest impact on the predictions are "MedInc","AveBedrms", "Latitude" and "Longitude". This is because their respective coefficients are ~0.49, ~0.78, ~0.42 and ~0.43, which have a larger magnitude relative to the other feature coefficients. This indicates that the features "MedInc","AveBedrms", "Latitude" and "Longitude" are more likely to have a positive linear relationship with the overall prediction of "med_house_value".
How well do the predicted values match the actual values?
The predicted values match the actual values moderately, since the R^2 score is 0.58. Since the standard deviation for our target feature is ~1.15 and the RMSE is as much as 0.75, this is another indication that the model is moderately good at predicting the actual values. 


In [45]:
#Build a simplified model using 3 select features
select_features = ["MedInc", "AveBedrms", "Longitude"]
X = X[select_features]

I chose the three features I chose for the simplified regression model because these features had the coefficients with the largest magnitudes. The features with coefficients with larger magnitudes have a greater influence on the prediction, and so in order to create the most accurate simplified model possible, it made sense to do so using the features that have the most direct relationship with the target feature.

In [47]:
#Train the simplified model using only these 3 features

#Split the dataset into training and test sets (80% training, 20% testing)
X_train_simp, X_test_simp, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Train the simplified linear regression model
lin_reg_simp = LinearRegression()
lin_reg_simp.fit(X_train_simp, y_train)
#Make predictions on the test set
y_pred_simp = lin_reg_simp.predict(X_test_simp)

In [ ]:

#Evaluate the performance of the simplified model using Mean Squared Error (MSE), Root Mean Squared Error (RMSE) and R^2 score
mse_simp = mean_squared_error(y_test, y_pred_simp)
rmse_simp = root_mean_squared_error(y_test, y_pred_simp)
r2_simp = r2_score(y_test, y_pred_simp)

print("Simplified Data Model Performance Evaluation: ")
print(f"Mean Squared Error: {mse_simp:.2f}")
print(f"Root Mean Squared Error: {rmse_simp:.2f}")
print(f"R² Score: {r2_simp:.2f}")

print("Full Data Model Performance Evaluation: ")
print(f"Mean Squared Error: {mse_raw:.2f}")
print(f"Root Mean Squared Error: {rmse_raw:.2f}")
print(f"R² Score: {r2_raw:.2f}")

Simplified Data Model Performance Evaluation: 
Mean Squared Error: 0.71
Root Mean Squared Error: 0.84
R² Score: 0.46
Full Data Model Performance Evaluation: 
Mean Squared Error: 0.56
Root Mean Squared Error: 0.75
R² Score: 0.58


Compared to the full model, the simplified model is less accurate. The MSE of the simplified model is higher than the full model by 0.15, while the RMSE of the simplified model is higher than the full model by 0.09. This means that the prediction errors are higher in the simplified model. In addition, the R^2 score for the simplified model is 0.12 lower, which means that variance between predictions is less accounted for by the model and that the simplified model is not able to capture as much variance as the full model. 
In practice, it would be preferable not to use a simplified model over a full model, since full models are more accurate. However, if the features you used to build the simplified model had coefficients with large magnitudes while the features you left out of the simplified model had coefficients with significanly smaller magnitudes, it would make sense to use a simplified model if it were more convenient to do so.